tests/usbus/main.c

```diff
@@ -0,0 +1,349 @@
+{
+    DEBUG("[validation]: IN ACK validation\n");
+
+    TEST_ASSERT_EQUAL_INT(ep->ep.dir, USB_EP_DIR_IN);
+    TEST_ASSERT_EQUAL_INT(len, 0);
+    dev->usbdev.epcb(&dev->in[0].ep, USBDEV_EVENT_ESR);
+    req_phase = TEST_REQ_PHASE_IDLE;
+}
+
+static void _validate_set_req_start(usbdev_mock_t *dev, usbdev_mock_ep_t *ep,
+                                    size_t len)
+{
+    DEBUG("[validation]: set request validation\n");
+
+    TEST_ASSERT_EQUAL_INT(dev->out[0].available, 8);
+    /* We should get an IN request with the data */
+    TEST_ASSERT_EQUAL_INT(ep->ep.dir, USB_EP_DIR_IN);
+    if (dev->req_len == 0) {
+        DEBUG("[validation]: No payload, performing IN ACK validation\n");
+        /* No data phase, this `ready` is the OUT ACK */
+        _validate_in_ack(dev, ep, len);
+    }
+    else {
+        req_phase = TEST_REQ_PHASE_OUTDATA;
+    }
+}
+
+static void _handle_data(usbdev_mock_t *dev, usbdev_mock_ep_t *ep, size_t len)
+{
+    if (ep->ep.dir == USB_EP_DIR_IN) {
+        /* Zero length IN packet is not valid here */
+        DEBUG("[data]: Handling IN data from stack: %u\n", (unsigned)len);
+        TEST_ASSERT(len != 0);
+        dev->usbdev.epcb(&dev->in[0].ep, USBDEV_EVENT_ESR);
+        if (ep->available == dev->req_len) {
+            DEBUG("[data]: Full data received from stack\n");
+            req_phase = TEST_REQ_PHASE_OUTACK;
+            /* Reset buffer ptr to the start */
+            ep->ep.buf = ep->buf_start;
+        }
+        else {
+            DEBUG("[data]: Expecting more data from stack: %u/%u\n",
+                  (unsigned)ep->available,
+                  (unsigned)dev->req_len);
+        }
+    }
+}
+
+static void _ep_esr_validation(usbdev_mock_t *dev, usbdev_mock_ep_t *ep)
+{
+    DEBUG("[ep esr]: Data available for stack: %u\n", ep->available);
+    if (req_phase == TEST_REQ_PHASE_IDLE) {
+        DEBUG("[ep esr]: Done with the request\n");
+        /* signal USBDEV_EVENT_ESR to call _test_sequence */
+        dev->usbdev.cb(&dev->usbdev, USBDEV_EVENT_ESR);
+    }
+    ep->ep.dev->epcb(&ep->ep, USBDEV_EVENT_TR_COMPLETE);
+}
+
+/* Handles Control endpoint validation from the usbdev side. *
+ * Called as part of the endpoint ready function.            */
+static void _ep0_validation(usbdev_mock_t *dev, usbdev_mock_ep_t *ep,
+                            size_t len)
+{
+    DEBUG("[validation]: req state: %s\n", _state2str(req_phase));
+    /* Validate direction and content for the current state */
+    switch (req_phase) {
+        case TEST_REQ_PHASE_START_GET:
+            _validate_get_req_start(ep);
+            _handle_data(dev, ep, len);
+            break;
+        case TEST_REQ_PHASE_START_SET:
+            _validate_set_req_start(dev, ep, len);
+            break;
+        case TEST_REQ_PHASE_OUTACK:
+            _validate_out_ack(dev, ep);
+            break;
+        case TEST_REQ_PHASE_INACK:
+            _validate_in_ack(dev, ep, len);
+            break;
+        default:
```

Comment:
please add a DEBUG statement

Reply:
The `default` case here is actively used with:
 - `TEST_REQ_PHASE_IDLE`
 - `TEST_REQ_PHASE_OUTDATA`
 - `TEST_REQ_PHASE_INDATA`

I don't assert anything at the moment for those stages.

I can add a comment there to clarify this or I can add a debug statement if you have a suggestion that is generic enough supply sensible information for these 3 cases.

Reply:
I can also split out these cases and add a simple debug statement for each, which is what I prefer as that would make the test easier to follow when `DEBUG` is enabled (I think).

Any preference?

Reply:
a comment should be enough, there is no need to spam output if it's heavily used.

Reply:
And added a comment :)